### PR TITLE
Update 05.model-complexity.md

### PR DESCRIPTION
**Pull request by jaybee84**
### Purpose

addresses #175 

#### Which areas should receive a particularly close look?
model_complexity.md has newer text to define ML related terms as well as provide examples before concepts. Please take a look to see if the draft achieved this. Refer to the linked issue to get more context on the changes.

#### Is there anything that you want to discuss further?
Are there things missing? Does it need more restructuring?

#### Is the pull request ready for review?
Yes

### Manuscript checklist

Unless otherwise noted above, this PR will be considered ready for review when all four items have been checked.

- [ ] All changes to text follow "one sentence per line" [[Manubot instructions](https://github.com/jaybee84/ml-in-rd/blob/draft-branch/USAGE.md#manubot-markdown)]
- [ ] All citations follow the [Manubot citation instructions](https://github.com/jaybee84/ml-in-rd/blob/draft-branch/USAGE.md#citations)
- [ ] All changes or additions to tables follow the [Manubot table instructions](https://github.com/jaybee84/ml-in-rd/blob/draft-branch/USAGE.md#tables)
- [ ] All figures follow the [Manubot figure instructions](https://github.com/jaybee84/ml-in-rd/blob/draft-branch/USAGE.md#figures)
- [ ] All terms that need to be added to glossary have been added to #50 
- [ ] There are no new spelling errors identified by the Manubot spellchecker

### CODE REVIEW

**Comment by jaybee84** [comment]
I had to merge in draft branch to resolve conflicts. Please take a look at the `06.model-complexity.md` file and you can ignore the rest. :)

**Comment by jaybee84** [comment]
oh no! I messed up the diff :(

**Comment by jaybee84** [comment]
pinging this PR to bump the review request up on everyone's inbox :)

File: content/06.model-complexity.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
+### Manage model complexity while preserving the value of machine learning
+
+
+<!-- TODO: stick to one example method of each approach so that we get to describe why we need to use it --->
+
+Machine learning methods are generally accompanied by a few critical assumptions. 
+Assumption 1, the dataset contains equal number of samples for all classes in a dataset (i.e. no imbalance among classes).
```
Comment (allaway) [inline content/06.model-complexity.md:7]: I would recommend making these three sentences start with `1.` `2.` etc if you really want a list, but I also (personally/stylistically) do not like lists :)

File: content/06.model-complexity.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
+### Manage model complexity while preserving the value of machine learning
+
+
+<!-- TODO: stick to one example method of each approach so that we get to describe why we need to use it --->
+
+Machine learning methods are generally accompanied by a few critical assumptions. 
+Assumption 1, the dataset contains equal number of samples for all classes in a dataset (i.e. no imbalance among classes). 
+Assumption 2, the dataset is complete with adequate samples for all features (lack of sparsity).
+Assumption 3 (for supervised learning methods), there is no ambiguity about the labels for the samples in the dataset (i.e. no label-noise).
+In reality, the rare disease datasets violate many of these assumptions.
```
Comment (allaway) [inline content/06.model-complexity.md]:
```suggestion
Rare disease datasets, however, violate many of these assumptions.
```

File: content/06.model-complexity.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+There is generally a high class imbalance due to small number of samples for specific classes, abundant label-noise due to incomplete understanding of the disease, and in many cases sparsity in data.
```
Comment (allaway) [inline content/06.model-complexity.md]:
```suggestion
There is generally a high class imbalance due to small number of samples for specific classes (e.g., only a few patients with a particular rare disease in a health records dataset), abundant label-noise due to incomplete understanding of the disease, and in many cases sparsity in data.
```

File: content/06.model-complexity.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+All of these contribute to low signal to noise ratio in rare disease datasets.
+Thus applying ML to rare disease data without any accomodations for the above shortcomings may lead to models that lack stability required for reproducibility or simplicity required for interpretation.
```
Comment (allaway) [inline content/06.model-complexity.md]:
```suggestion
Thus, applying ML to rare disease data without any accomodations for the above shortcomings may lead to models that lack stability required for reproducibility or simplicity required for interpretation.
```

File: content/06.model-complexity.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+In this section we highlight a few common ML techniques that can accomodate the three above mentioned aspects and improve the stability and simplicity of ML models applied to rare disease data.
```
Comment (allaway) [inline content/06.model-complexity.md]:
```suggestion
In this section, we highlight a few ML techniques that can help manage the challenges in applying ML models applied to rare disease data.
```

File: content/06.model-complexity.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+Stability in predictions while accommodating class imbalance in datasets can be achieved through decision tree based ensemble learning methods (Figure[@fig:2]A-B).
```
Comment (allaway) [inline content/06.model-complexity.md]:
```suggestion
Accommodating class imbalance in datasets and increasing the stability of ML predictions can be achieved through decision tree-based ensemble learning methods (Figure[@fig:2]A-B).
```

File: content/06.model-complexity.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+Random forests use resampling (with replacement) based techniques like _bagging_ (bootstrap aggregation) of independent decision trees to form a consensus about the important predictive features [@https://doi.org/10.1023/A:1010933404324; @doi:10.1186/1472-6947-13-134; @doi:10.1214/aos/1031689014; @doi:10.1177/2045894019890549; @doi:10/btzfh6]. 
+Additional approaches like combining random forests with resampling without replacement can generate confidence intervals for the model predictions by iteratively exposing the models to incomplete datasets, mimicking real world cases where most rare disease datasets are incomplete [@doi:10.3390/genes11020226].
+But recent studies suggest that there are limitations to decision tree-based ensemble methods when applied to rare disease datasets with substantial class imbalance and label-noise [@doi:10.1007/s11634-019-00354-x; @pmid:30815073].
```
Comment (allaway) [inline content/06.model-complexity.md]:
```suggestion
Recent studies suggest that there are limitations to decision tree-based ensemble methods when applied to rare disease datasets with substantial class imbalance and label-noise [@doi:10.1007/s11634-019-00354-x; @pmid:30815073]. 
```

File: content/06.model-complexity.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+This has led to the adoption of cascade learning, a variant of ensemble learning, where multiple methods leveraging distinct underlying assumptions are used in tandem; and augmented with algorithms like AdaBoost (boosting) to capture stable patterns existing in silver standard data [@doi:10.1109/cvpr.2001.990537; @doi:10.1007/978-3-540-75175-5_16; @doi:10.1109/icpr.2004.1334680]. 
+For example, a cascade learning approach for identifying rare disease patients from electronic health record data utilized independent steps for feature extraction (word2vec [@arxiv:1301.3781]), preliminary prediction with ensembled decision trees, and prediction refinement using data similarity metrics [@pmid:30815073]. 
+Combining these three methods resulted in better performance than other methods when implemented on the silver standard dataset in isolation.
+Other approaches like inverse class weighting or oversampling (SMOTE) or cascade learning approaches that adequately represent the rarer classes are also beneficial in increasing stability of predictions while accommodating for class imbalance.
```
Comment (allaway) [inline content/06.model-complexity.md]:
```suggestion
Other approaches like inverse class weighting, oversampling (with methods like Synthetic Minority Oversampling Technique, abbreviated SMOTE), or cascade learning approaches that adequately represent the rarer classes are also beneficial in increasing stability of predictions while accommodating for class imbalance.
```

File: content/06.model-complexity.md [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+Presence of label-noise and sparsity in the data can lead to overfitting of models to the training data, such that the models show high prediction accuracy on training data but low prediction accuracy on encountering new data points. 
+Thus features identified as predictive by such overfit models are often found to be tailored to the training dataset making the model overtly complex, and not generalizable to new data points. [@doi:10.1073/pnas.1900654116]
```
Comment (allaway) [inline content/06.model-complexity.md]:
```suggestion
Features identified as predictive by overfit models are often mis-represented in the training dataset making the model not generalizable to new data. [@doi:10.1073/pnas.1900654116]
```

File: content/06.model-complexity.md [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+Regularization is often used in rare variant discovery and immune cell signature discovery studies which are useful examples that also need to accomodate sparsity of data like rare diseases.
+For example, among various regularization methods (ridge regression, LASSO regression, and elastic-net)[@doi:10.1111/j.1467-9868.2005.00503.x], hybrid applications of LASSO for capturing combinations of rare variants have proven beneficial [@doi:10.1016/j.ajhg.2008.06.024; @doi:10.1186/1753-6561-5-s9-s113] .
+In the context of rare immune cell signature discovery, elastic-net regression was found to outperform other regression approaches [@doi:10.1111/j.1467-9868.2005.00503.x; @doi:10.1016/j.compbiomed.2015.10.008; @doi:10.1186/1471-2105-14-198; @doi:10.1186/s12859-019-2994-z]. 
+Thus regularization methods like LASSO or elastic-net have been methods of choice while working with rare observations; use of these regularization approaches should be considered while working with rare disease datasets. 
+
```
Comment (allaway) [inline content/06.model-complexity.md:32]: I am wondering if the section above has too much jargon. What do you  all think? If so, maybe this could be dealt with in another PR.

File: content/07.prior-knowledge.md [hunk 11]
```diff
@@ -1,14 +1,24 @@
...
+Rare disease researchers can repurpose general (i.e., not rare disease-specific) biological or chemical knowledge graphs to answer rare disease-based research questions [@doi:10.1142/9789811215636_0041]. 
+There are a variety of tactics to sift through the large amounts of complex data in knowledge graphs.
+One such tactic is to calculate the distances between nodes of interest (e.g., diseases and drugs to identify drugs for repurposing in rare disease [@doi:10.1142/9789811215636_0041]); this is often done by determining the "embeddings" (linear representations of the position and connections of a particular point in the graph) for nodes in the knowledge graph, and calculating the similarity between these embeddings. 
+Testing a variety of methods different methods to calculate node embeddings that can generate actionable insights for rare diseases is an active area of research [@doi:10.1142/9789811215636_0041], and an opportunity for continued research.
+Another application of KGs that has been previously described is to augment or refine a dataset [@doi:10.1186/s12911-019-0752-9, doi:10.1186/s12911-019-0938-1].
```
Comment (allaway) [inline content/07.prior-knowledge.md]:
```suggestion
Another application of KGs is to augment or refine a dataset [@doi:10.1186/s12911-019-0752-9, doi:10.1186/s12911-019-0938-1].
```

File: content/07.prior-knowledge.md [hunk 12]
```diff
@@ -1,14 +1,24 @@
...
+For example, Li, et. al.[@doi:10.1186/s12911-019-0938-1] used a KG to identify linked terms in a medical corpus from a large number of patients, some with rare disease diagnoses .
```
Comment (allaway) [inline content/07.prior-knowledge.md]:
```suggestion
For example, Li, et. al.[@doi:10.1186/s12911-019-0938-1] used a KG to identify linked terms in a medical corpus from a large number of patients, some with rare disease diagnoses.
```

File: content/07.prior-knowledge.md [hunk 13]
```diff
@@ -1,14 +1,24 @@
...
+They were able to augment their text dataset by identifying related terms in the clinical text to map them to the same term - an example given being mapping "cancer" and "malignancy" in different patients to the same clinical concept.
```
Comment (allaway) [inline content/07.prior-knowledge.md]:
```suggestion
They were able to augment their text dataset by identifying related terms in the clinical text to map them to the same term - e.g., mapping "cancer" and "malignancy" in different patients to the same clinical concept. 
```

File: content/07.prior-knowledge.md [hunk 14]
```diff
@@ -1,14 +1,24 @@
...
+Another application of KGs that has been previously described is to augment or refine a dataset [@doi:10.1186/s12911-019-0752-9, doi:10.1186/s12911-019-0938-1].
+For example, Li, et. al.[@doi:10.1186/s12911-019-0938-1] used a KG to identify linked terms in a medical corpus from a large number of patients, some with rare disease diagnoses .
+They were able to augment their text dataset by identifying related terms in the clinical text to map them to the same term - an example given being mapping "cancer" and "malignancy" in different patients to the same clinical concept. 
+With this augmented and improved dataset, they were able to train and test a variety of text classification algorithms to identify rare disease patients within their corpus. 
+Finally, another possible tactic for rare disease researchers is to take a knowledge graphs (or an integration of several knowledge graphs) and apply neural network-based algorithms optimized for graph data, such as a graph convolutional neural network.
```
Comment (allaway) [inline content/07.prior-knowledge.md]:
```suggestion
Finally, another possible tactic for rare disease researchers is to take a knowledge graph (or an integration of several knowledge graphs) and apply neural network-based algorithms optimized for graph data, such as a graph convolutional neural network.
```

File: content/07.prior-knowledge.md [hunk 15]
```diff
@@ -1,14 +1,24 @@
...
+Rao and colleagues [@doi:10.1186/s12920-018-0372-8] describe the construction of a KG using phenotype information (Human Phenotype Ontology) and rare disease information (Orphanet) and curated gene interaction/pathway data (Lit-BM-13, Wikipathways). 
+They then trained a spectral graph convolution neural network on this knowledge graph to identify and rank potentially causal genes for the Orphanet rare diseases, and were able to use this model to predict causal genes for a ground truth dataset of rare diseases with known causal genes.
```
Comment (allaway) [inline content/07.prior-knowledge.md]:
```suggestion
They then trained an algorithm (a spectral graph convolution neural network) on this knowledge graph to identify and rank potentially causal genes for the Orphanet rare diseases, and were able to use this model to predict causal genes for a ground truth dataset of rare diseases with known causal genes. 
```
Not sure if this is necessary but might help folks read thru this section w/o knowing what a spectral gcn is

File: content/07.prior-knowledge.md [hunk 16]
```diff
@@ -1,14 +1,24 @@
...
+With this augmented and improved dataset, they were able to train and test a variety of text classification algorithms to identify rare disease patients within their corpus. 
+Finally, another possible tactic for rare disease researchers is to take a knowledge graphs (or an integration of several knowledge graphs) and apply neural network-based algorithms optimized for graph data, such as a graph convolutional neural network.
+Rao and colleagues [@doi:10.1186/s12920-018-0372-8] describe the construction of a KG using phenotype information (Human Phenotype Ontology) and rare disease information (Orphanet) and curated gene interaction/pathway data (Lit-BM-13, Wikipathways). 
+They then trained a spectral graph convolution neural network on this knowledge graph to identify and rank potentially causal genes for the Orphanet rare diseases, and were able to use this model to predict causal genes for a ground truth dataset of rare diseases with known causal genes. 
+While several groups have published on the use of KGs to study rare disease already, we expect that the increasing availibility of multi-modal data resources will make this a popular area of research in the future.
```
Comment (allaway) [inline content/07.prior-knowledge.md]:
```suggestion
While several groups have already published on the use of KGs to study rare diseases, we expect that the growth of multi-modal datasets and methods to analyze KGs will make them a more popular and important tool in the application of ML in rare disease. 
```

**Review by allaway — CHANGES_REQUESTED** [review]
Still need to review `content/04. box_1_experimental_design.md`

**Comment by jaybee84** [comment]
> Still need to review `content/04. box_1_experimental_design.md`

@allaway : the box_1_experimental_design was reviewed and merged in an earlier PR.. so you can skip it if you dont have the time. Thanks for reviewing this one... will make the necessary changes.

File: content/04. box_1_experimental_design.md [hunk 17]
```diff
@@ -0,0 +1,40 @@
+#### Box 1: Understanding experimental design of machine learning to inform requirements for data
+
+##### Components of ML experiments
+
+Machine learning algorithms identify patterns that explain or fit a given dataset.
+Every machine learning algorithm (_model_) goes through a _training_ phase where it identifies underlying patterns in a given dataset, and a _testing_ phase where the model applies the identified patterns to unseen data points.
```
Comment (allaway) [inline content/04. box_1_experimental_design.md]:
```suggestion
Every machine learning algorithm goes through a _training_ phase where it identifies underlying patterns in a given dataset to create a "trained" algorithm (a _model_) , and a _testing_ phase where the model applies the identified patterns to unseen data points.
```

I'd consider the "model" to be the trained predictor, not the naive algorithm, what do you all think?

File: content/04. box_1_experimental_design.md [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+So every model should be provided with the following fundamental parts as input: 1. a _training dataset_ , 2. an _evaluation dataset_ , 3. a _held-out validation dataset_ .
```
Comment (allaway) [inline content/04. box_1_experimental_design.md]:
```suggestion
Typically, a machine learning algorithm is provided with the following fundamental parts as input: 1. a _training dataset_ , 2. an _evaluation dataset_ , 3. a _held-out validation dataset_ .
```

File: content/04. box_1_experimental_design.md [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+Typically each of these input datasets consists of a matrix (or dataframe or table) where samples form the rows and features form the columns.
```
Comment (allaway) [inline content/04. box_1_experimental_design.md]:
```suggestion
These input data can be images, text, numbers, or other types of data which are typically encoded as a numerical representation of the input data. 
```

I think this is safe to say is typically true, but there are probably exceptions....

File: content/04. box_1_experimental_design.md [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+A _training dataset_ is used to expose the model to underlying patterns among the features present in the data of interest.
+An _evaluation dataset_ is a small test dataset which is used during the training phase to help the model iteratively update its parameters (i.e., _hyperparameter_tuning_ or _model tuning_).
+In many cases, a large training set may be subdivided to form a smaller training dataset and the evaluation dataset and both used to train the model (see next section for more details on cross-validation).
```
Comment (allaway) [inline content/04. box_1_experimental_design.md]:
```suggestion
In many cases, a large training set may be subdivided to form a smaller training dataset and the evaluation dataset and both used to train the model (see next section for more details on cross-validation).
```
I think this sentence is a bit confusing but not sure how to correct it.

File: content/04. box_1_experimental_design.md [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+In the testing phase, a new or unseen test dataset or _held-out validation set_ is used to test whether the patterns learned by the model can be generalized to new data points.
```
Comment (allaway) [inline content/04. box_1_experimental_design.md]:
```suggestion
In the testing phase, a new or unseen test dataset or _held-out validation set_ is used to test whether the patterns learned by the model hold true in new data (are _generalizable_).
```

File: content/04. box_1_experimental_design.md [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+While the evaluation dataset helps the model iteratively update its parameters to learn important patterns in the training data, the held-out validation set helps test the generalizability of the model.
+Generalizability of a model is its ability to recognize patterns that can help predict the class or an outcome for previously unseen data.
```
Comment (allaway) [inline content/04. box_1_experimental_design.md:14]: This and the next few lines feels like it should be at the end, since this is chronologically at the end of the train-test cycle

File: content/04. box_1_experimental_design.md [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+This makes sure that all the datasets involved in training and testing a model maintain uniformity in the features.
+In case of rare diseases where multiple datasets may be combined to make a large enough training dataset, special care is taken to standardize the features and the patterns therein.
+The iterative training stage helps the model learn important patterns in the training dataset and then use the evaluation dataset to test for errors in prediction and update its learning parameters (_hyperparameter tuning_).
+The method by which the evaluation dataset tests the performance of the trained model and helps update the hyperparameters is called _cross-validation_.
+To maximally utilize the available data for cross-validation, there can be multiple approaches to form the training and evaluation datasets e.g. leave-p-out cross-validation, leave-one-out cross-validation, k-fold cross-validation, Monte-Carlo random subsampling cross-validation [@doi:10.1007/978-1-4614-6849-3].
+In case of k-fold cross-validation, a given dataset is shuffled randomly and split into _k_ parts.
```
Comment (allaway) [inline content/04. box_1_experimental_design.md:29]: Do we call this out because it's in the figure? or some other reason

File: content/06.model-complexity.md (hunk 1 above)
Comment (jaybee84) [reply content/06.model-complexity.md:7]:
```suggestion
1. the dataset contains equal number of samples for all classes in a dataset (i.e. no imbalance among classes). 
```

File: content/06.model-complexity.md [hunk 24]
```diff
@@ -0,0 +1,34 @@
+### Manage model complexity while preserving the value of machine learning
+
+
+<!-- TODO: stick to one example method of each approach so that we get to describe why we need to use it --->
+
+Machine learning methods are generally accompanied by a few critical assumptions. 
+Assumption 1, the dataset contains equal number of samples for all classes in a dataset (i.e. no imbalance among classes). 
+Assumption 2, the dataset is complete with adequate samples for all features (lack of sparsity).
```
Comment (jaybee84) [inline content/06.model-complexity.md]:
```suggestion
2. the dataset is complete with adequate samples for all features (lack of sparsity).
```

File: content/06.model-complexity.md [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+Assumption 3 (for supervised learning methods), there is no ambiguity about the labels for the samples in the dataset (i.e. no label-noise).
```
Comment (jaybee84) [inline content/06.model-complexity.md]:
```suggestion
3. there is no ambiguity about the labels for the samples in the dataset (i.e. no label-noise, applicable for supervised learning methods).
```

File: content/04. box_1_experimental_design.md (hunk 20 above)
Comment (jaybee84) [reply content/04. box_1_experimental_design.md]:
```suggestion
In many cases, a large training set may be subdivided to form a smaller training dataset and the evaluation dataset, both of which are used to train the model (see next section for more details on cross-validation).
```

File: content/04. box_1_experimental_design.md (hunk 23 above)
Comment (jaybee84) [reply content/04. box_1_experimental_design.md:29]: not quite sure what you are referring to here.. k-fold cross validation?

File: content/06.model-complexity.md (hunk 10 above)
Comment (jaybee84) [reply content/06.model-complexity.md:32]: will do..

File: content/04. box_1_experimental_design.md (hunk 23 above)
Comment (jaybee84) [reply content/04. box_1_experimental_design.md:29]: we used k-fold cross validation as an example in the text (since it is the easiest to explain :).. and then added this to the figure for illustrative purposes
